File: web/src/utils/validation.ts
/**
 * Validation Utility Functions for pxBlipCreator
 * Validates blip data and related inputs
 */

// Map bounds for GTA V (approximate)
const MAP_BOUNDS = {
  minX: -3000,
  maxX: 3000,
  minY: -3000,
  maxY: 3000,
  minZ: -100,
  maxZ: 1000,
};

// Valid FiveM blip sprites (0-1000+)
const VALID_SPRITE_MIN = 0;
const VALID_SPRITE_MAX = 1000;

// Valid color range
const VALID_COLOR_MIN = 0;
const VALID_COLOR_MAX = 255;

// Valid alpha (transparency) range
const VALID_ALPHA_MIN = 0;
const VALID_ALPHA_MAX = 255;

// Valid scale range (stored as size * 10)
const VALID_SCALE_MIN = 10;
const VALID_SCALE_MAX = 500;

export interface ValidationError {
  field: string;
  message: string;
}

export interface ValidationResult {
  isValid: boolean;
  errors: ValidationError[];
}

/**
 * Validates a blip name
 * @param name - Blip name to validate
 * @returns Validation errors (empty if valid)
 */
export const validateBlipName = (name: string | undefined): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (!name) {
    errors.push({ field: 'name', message: 'Blip name is required' });
    return errors;
  }

  // Trim and check length
  const trimmedName = name.trim();
  if (trimmedName.length === 0) {
    errors.push({ field: 'name', message: 'Blip name cannot be empty' });
  } else if (trimmedName.length > 50) {
    errors.push({ field: 'name', message: 'Blip name cannot exceed 50 characters' });
  }

  return errors;
};

/**
 * Validates coordinates
 * @param coords - Object with x, y, z properties
 * @returns Validation errors (empty if valid)
 */
export const validateCoords = (
  coords: { x?: number; y?: number; z?: number } | undefined
): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (!coords) {
    errors.push({ field: 'coords', message: 'Coordinates are required' });
    return errors;
  }

  // Validate X coordinate
  if (coords.x === undefined || coords.x === null) {
    errors.push({ field: 'coordsX', message: 'X coordinate is required' });
  } else if (typeof coords.x !== 'number') {
    errors.push({ field: 'coordsX', message: 'X must be a number' });
  } else if (coords.x < MAP_BOUNDS.minX || coords.x > MAP_BOUNDS.maxX) {
    errors.push({
      field: 'coordsX',
      message: `X coordinate must be between ${MAP_BOUNDS.minX} and ${MAP_BOUNDS.maxX}`,
    });
  }

  // Validate Y coordinate
  if (coords.y === undefined || coords.y === null) {
    errors.push({ field: 'coordsY', message: 'Y coordinate is required' });
  } else if (typeof coords.y !== 'number') {
    errors.push({ field: 'coordsY', message: 'Y must be a number' });
  } else if (coords.y < MAP_BOUNDS.minY || coords.y > MAP_BOUNDS.maxY) {
    errors.push({
      field: 'coordsY',
      message: `Y coordinate must be between ${MAP_BOUNDS.minY} and ${MAP_BOUNDS.maxY}`,
    });
  }

  // Validate Z coordinate
  if (coords.z === undefined || coords.z === null) {
    errors.push({ field: 'coordsZ', message: 'Z coordinate is required' });
  } else if (typeof coords.z !== 'number') {
    errors.push({ field: 'coordsZ', message: 'Z must be a number' });
  } else if (coords.z < MAP_BOUNDS.minZ || coords.z > MAP_BOUNDS.maxZ) {
    errors.push({
      field: 'coordsZ',
      message: `Z coordinate must be between ${MAP_BOUNDS.minZ} and ${MAP_BOUNDS.maxZ}`,
    });
  }

  return errors;
};

/**
 * Validates sprite ID
 * @param sprite - Sprite ID to validate
 * @returns Validation errors (empty if valid)
 */
export const validateSprite = (sprite: number | undefined): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (sprite === undefined || sprite === null) {
    errors.push({ field: 'sprite', message: 'Sprite is required' });
    return errors;
  }

  if (typeof sprite !== 'number' || !Number.isInteger(sprite)) {
    errors.push({ field: 'sprite', message: 'Sprite must be an integer' });
  } else if (sprite < VALID_SPRITE_MIN || sprite > VALID_SPRITE_MAX) {
    errors.push({
      field: 'sprite',
      message: `Sprite must be between ${VALID_SPRITE_MIN} and ${VALID_SPRITE_MAX}`,
    });
  }

  return errors;
};

/**
 * Validates color value
 * @param color - Color value to validate
 * @param fieldName - Field name for error message
 * @returns Validation errors (empty if valid)
 */
export const validateColor = (
  color: number | undefined,
  fieldName: string = 'color'
): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (color === undefined || color === null) {
    errors.push({ field: fieldName, message: `${fieldName} is required` });
    return errors;
  }

  if (typeof color !== 'number' || !Number.isInteger(color)) {
    errors.push({ field: fieldName, message: `${fieldName} must be an integer` });
  } else if (color < VALID_COLOR_MIN || color > VALID_COLOR_MAX) {
    errors.push({
      field: fieldName,
      message: `${fieldName} must be between ${VALID_COLOR_MIN} and ${VALID_COLOR_MAX}`,
    });
  }

  return errors;
};

/**
 * Validates alpha (transparency)
 * @param alpha - Alpha value to validate
 * @returns Validation errors (empty if valid)
 */
export const validateAlpha = (alpha: number | undefined): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (alpha === undefined || alpha === null) {
    errors.push({ field: 'alpha', message: 'Alpha is required' });
    return errors;
  }

  if (typeof alpha !== 'number' || !Number.isInteger(alpha)) {
    errors.push({ field: 'alpha', message: 'Alpha must be an integer' });
  } else if (alpha < VALID_ALPHA_MIN || alpha > VALID_ALPHA_MAX) {
    errors.push({
      field: 'alpha',
      message: `Alpha must be between ${VALID_ALPHA_MIN} and ${VALID_ALPHA_MAX}`,
    });
  }

  return errors;
};

/**
 * Validates scale
 * @param scale - Scale value to validate
 * @returns Validation errors (empty if valid)
 */
export const validateScale = (scale: number | undefined): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (scale === undefined || scale === null) {
    errors.push({ field: 'scale', message: 'Scale is required' });
    return errors;
  }

  if (typeof scale !== 'number') {
    errors.push({ field: 'scale', message: 'Scale must be a number' });
  } else if (scale < VALID_SCALE_MIN || scale > VALID_SCALE_MAX) {
    errors.push({
      field: 'scale',
      message: `Scale must be between ${VALID_SCALE_MIN} and ${VALID_SCALE_MAX}`,
    });
  }

  return errors;
};

/**
 * Sanitizes string for safe display and storage
 * Removes potentially dangerous characters but keeps most input
 * @param str - String to sanitize
 * @returns Sanitized string
 */
export const sanitizeString = (str: string | undefined): string => {
  if (!str) return '';

  // Trim whitespace
  let sanitized = str.trim();

  // Remove null bytes
  sanitized = sanitized.replace(/\0/g, '');

  // Limit length to database field max
  if (sanitized.length > 50) {
    sanitized = sanitized.substring(0, 50);
  }

  return sanitized;
};

/**
 * Validates a complete blip object
 * @param blip - Blip object to validate
 * @returns ValidationResult with all errors
 */
export const validateBlipData = (blip: any): ValidationResult => {
  const errors: ValidationError[] = [];

  if (!blip) {
    errors.push({ field: 'blip', message: 'Blip data is required' });
    return { isValid: false, errors };
  }

  // Validate each field
  errors.push(...validateBlipName(blip.name));
  errors.push(...validateCoords(blip.coords));
  errors.push(...validateSprite(blip.Sprite));
  errors.push(...validateColor(blip.sColor, 'sColor'));
  errors.push(...validateAlpha(blip.alpha));
  errors.push(...validateScale(blip.scale));

  return {
    isValid: errors.length === 0,
    errors,
  };
};

/**
 * Validates all RGB color values in an array
 * @param colors - Array with [R, G, B] values
 * @returns Validation errors (empty if valid)
 */
export const validateRGBArray = (colors: number[] | undefined): ValidationError[] => {
  const errors: ValidationError[] = [];

  if (!colors) {
    errors.push({ field: 'colors', message: 'Colors array is required' });
    return errors;
  }

  if (!Array.isArray(colors) || colors.length !== 3) {
    errors.push({ field: 'colors', message: 'Colors must be an array of [R, G, B]' });
    return errors;
  }

  // Validate each color component
  colors.forEach((color, index) => {
    if (typeof color !== 'number' || !Number.isInteger(color)) {
      errors.push({
        field: `colors[${index}]`,
        message: `Color component at index ${index} must be an integer`,
      });
    } else if (color < VALID_COLOR_MIN || color > VALID_COLOR_MAX) {
      errors.push({
        field: `colors[${index}]`,
        message: `Color component must be between ${VALID_COLOR_MIN} and ${VALID_COLOR_MAX}`,
      });
    }
  });

  return errors;
};

export default {
  validateBlipName,
  validateCoords,
  validateSprite,
  validateColor,
  validateAlpha,
  validateScale,
  sanitizeString,
  validateBlipData,
  validateRGBArray,
};

File: server/validation.lua
local MAP_BOUNDS = {
	minX = -3000,
	maxX = 3000,
	minY = -3000,
	maxY = 3000,
	minZ = -100,
	maxZ = 1000,
}

-- Valid FiveM blip sprites range
local VALID_SPRITE_MIN = 0
local VALID_SPRITE_MAX = 1000

-- Valid color range
local VALID_COLOR_MIN = 0
local VALID_COLOR_MAX = 255

-- Valid alpha (transparency) range
local VALID_ALPHA_MIN = 0
local VALID_ALPHA_MAX = 255

-- Maximum blip name length
local MAX_NAME_LENGTH = 50

--- Trims whitespace from a string
local function trim(str)
	return str:match("^%s*(.-)%s*$")
end

--- Checks if a value is a valid number
local function isValidNumber(value)
	return type(value) == "number" and not (value ~= value) -- NaN check
end

--- Validates a blip name
-- @param name The blip name to validate
-- @return Boolean indicating validity, error message if invalid
local function validateName(name)
	if not name or type(name) ~= "string" then
		return false, "Blip name must be a string"
	end

	local trimmed = trim(name)
	if #trimmed == 0 then
		return false, "Blip name cannot be empty"
	end

	if #trimmed > MAX_NAME_LENGTH then
		return false, string.format("Blip name cannot exceed %d characters", MAX_NAME_LENGTH)
	end

	return true, trimmed
end

--- Validates coordinates
-- @param coords Table with x, y, z properties
-- @return Boolean indicating validity, coords if valid or error message if invalid
local function validateCoords(coords)
	if not coords or type(coords) ~= "table" then
		return false, "Coordinates must be a table"
	end

	local x, y, z = coords.x, coords.y, coords.z

	-- Validate X
	if not isValidNumber(x) then
		return false, "X coordinate must be a valid number"
	end
	if x < MAP_BOUNDS.minX or x > MAP_BOUNDS.maxX then
		return false, string.format("X coordinate must be between %.0f and %.0f", 
			MAP_BOUNDS.minX, MAP_BOUNDS.maxX)
	end

	-- Validate Y
	if not isValidNumber(y) then
		return false, "Y coordinate must be a valid number"
	end
	if y < MAP_BOUNDS.minY or y > MAP_BOUNDS.maxY then
		return false, string.format("Y coordinate must be between %.0f and %.0f",
			MAP_BOUNDS.minY, MAP_BOUNDS.maxY)
	end

	-- Validate Z
	if not isValidNumber(z) then
		return false, "Z coordinate must be a valid number"
	end
	if z < MAP_BOUNDS.minZ or z > MAP_BOUNDS.maxZ then
		return false, string.format("Z coordinate must be between %.0f and %.0f",
			MAP_BOUNDS.minZ, MAP_BOUNDS.maxZ)
	end

	return true, { x = x, y = y, z = z }
end

--- Validates sprite ID
-- @param sprite The sprite ID to validate
-- @return Boolean indicating validity, sprite if valid or error message if invalid
local function validateSprite(sprite)
	if not isValidNumber(sprite) then
		return false, "Sprite must be a valid number"
	end

	if sprite ~= math.floor(sprite) then
		return false, "Sprite must be an integer"
	end

	if sprite < VALID_SPRITE_MIN or sprite > VALID_SPRITE_MAX then
		return false, string.format("Sprite must be between %d and %d",
			VALID_SPRITE_MIN, VALID_SPRITE_MAX)
	end

	return true, sprite
end

--- Validates a color value
-- @param color The color value to validate
-- @param fieldName Optional field name for error message
-- @return Boolean indicating validity, color if valid or error message if invalid
local function validateColor(color, fieldName)
	fieldName = fieldName or "color"

	if not isValidNumber(color) then
		return false, string.format("%s must be a valid number", fieldName)
	end

	if color ~= math.floor(color) then
		return false, string.format("%s must be an integer", fieldName)
	end

	if color < VALID_COLOR_MIN or color > VALID_COLOR_MAX then
		return false, string.format("%s must be between %d and %d",
			fieldName, VALID_COLOR_MIN, VALID_COLOR_MAX)
	end

	return true, color
end

--- Validates alpha (transparency) value
-- @param alpha The alpha value to validate
-- @return Boolean indicating validity, alpha if valid or error message if invalid
local function validateAlpha(alpha)
	if not isValidNumber(alpha) then
		return false, "Alpha must be a valid number"
	end

	if alpha ~= math.floor(alpha) then
		return false, "Alpha must be an integer"
	end

	if alpha < VALID_ALPHA_MIN or alpha > VALID_ALPHA_MAX then
		return false, string.format("Alpha must be between %d and %d",
			VALID_ALPHA_MIN, VALID_ALPHA_MAX)
	end

	return true, alpha
end

--- Validates scale value
-- @param scale The scale value to validate
-- @return Boolean indicating validity, scale if valid or error message if invalid
local function validateScale(scale)
	if not isValidNumber(scale) then
		return false, "Scale must be a valid number"
	end

	-- Allow reasonable scale values (stored as size * 10)
	if scale < 10 or scale > 500 then
		return false, "Scale must be between 10 and 500"
	end

	return true, scale
end

--- Sanitizes a string for safe storage
-- @param str The string to sanitize
-- @return Sanitized string
local function sanitizeString(str)
	if not str or type(str) ~= "string" then
		return ""
	end

	-- Trim whitespace
	local sanitized = trim(str)

	-- Remove null bytes
	sanitized = sanitized:gsub("%z", "")

	-- Limit to database field max
	if #sanitized > MAX_NAME_LENGTH then
		sanitized = sanitized:sub(1, MAX_NAME_LENGTH)
	end

	return sanitized
end

--- Validates RGB color array
-- @param colors Array with [R, G, B] values
-- @return Boolean indicating validity, colors if valid or error message if invalid
local function validateRGBArray(colors)
	if not colors or type(colors) ~= "table" then
		return false, "Colors must be a table"
	end

	if #colors ~= 3 then
		return false, "Colors must have exactly 3 values [R, G, B]"
	end

	-- Validate each color component
	for i = 1, 3 do
		local valid, color = validateColor(colors[i], string.format("Color[%d]", i))
		if not valid then
			return false, color
		end
	end

	return true, colors
end

--- Validates an entire blip object
-- @param blip The blip table to validate
-- @return Boolean indicating validity, error message if invalid, or sanitized blip if valid
local function validateBlip(blip)
	if not blip or type(blip) ~= "table" then
		return false, "Blip must be a table"
	end

	-- Validate name
	local nameValid, name = validateName(blip.name)
	if not nameValid then
		return false, name
	end

	-- Validate coordinates if provided
	if blip.coords then
		local coordsValid, coords = validateCoords(blip.coords)
		if not coordsValid then
			return false, coords
		end
		blip.coords = coords
	end

	-- Validate sprite if provided
	if blip.Sprite then
		local spriteValid, sprite = validateSprite(blip.Sprite)
		if not spriteValid then
			return false, sprite
		end
		blip.Sprite = sprite
	end

	-- Validate colors if provided
	if blip.sColor then
		local colorValid, color = validateColor(blip.sColor, "sColor")
		if not colorValid then
			return false, color
		end
		blip.sColor = color
	end

	-- Validate alpha if provided
	if blip.alpha then
		local alphaValid, alpha = validateAlpha(blip.alpha)
		if not alphaValid then
			return false, alpha
		end
		blip.alpha = alpha
	end

	-- Validate scale if provided
	if blip.scale then
		local scaleValid, scale = validateScale(blip.scale)
		if not scaleValid then
			return false, scale
		end
		blip.scale = scale
	end

	-- Sanitize string fields
	if blip.name then
		blip.name = sanitizeString(blip.name)
	end

	return true, blip
end

return {
	validateName = validateName,
	validateCoords = validateCoords,
	validateSprite = validateSprite,
	validateColor = validateColor,
	validateAlpha = validateAlpha,
	validateScale = validateScale,
	sanitizeString = sanitizeString,
	validateRGBArray = validateRGBArray,
	validateBlip = validateBlip,
	MAP_BOUNDS = MAP_BOUNDS,
	MAX_NAME_LENGTH = MAX_NAME_LENGTH,
}
